#!/usr/bin/env Rscript

# Function that takes an alpha_diversity table
# generated by RTK and generates various plots.
# National Research Council Canada
# Genomics and Microbiomes
# Author Julien Tremblay - julien.tremblay@nrc-cnrc.gc.ca
generateAlphaDivPlots <- function(infile, outdir, prefix, mappingFile) {
   
   options(stringsAsFactors = FALSE)
   library(ggplot2)
   library(grid)
   library(scales)
   library(gridExtra)
   library(data.table)
   .e <- environment()

   #infile = "~/Projects/R_debug/alpha_richness.tsv"
   #mappingFile = "~/Projects/R_debug/mapping_file.tsv"
   #outdir = "~/Projects/R_debug/"
   #prefix = "observedFeatures"
   read_depth = as.character(1000)
   
   alpha_div = data.frame(fread(infile, sep="\t", header=T, fill=TRUE), check.names=FALSE)
   alpha_div = alpha_div[!alpha_div$depth %in% c("taxonomy"),]
   alpha_div = alpha_div[,colnames(alpha_div) %in% c("depth", read_depth)]
   alpha_div = data.frame(t(alpha_div))
   colnames(alpha_div) = alpha_div[1,]
   alpha_div = alpha_div[-1,]
   alpha_div[,1] = NULL
   
   mapping_file = read.table(mappingFile, sep="\t", colClasses = "character")
   header = read.table(mappingFile, sep="\t", nrows=1, comment.char="")
   colnames(mapping_file) = header

   for(i in 2:length(header)){

      # Remove groups only having one variable - will crash otherwise
      if(length(unique(mapping_file[[ header[,i] ]])) < 2){
         next;
      } 
    
      # Remove all variable with less than 3 variable for each group, 
      df = data.frame(table(mapping_file[[ header[,i] ]]))
      df = df[df$Freq>=3,]
      if(nrow(df)<2){
         print(paste0("Skipping ", header[i], " because less than 2 reps per variable..."))
         next;
      }
      
      
      sample_labels = colnames(alpha_div)
      mapping_file_parsed = mapping_file[mapping_file[,1] %in% sample_labels,]
  
      colors <- c('#00FF00', '#FF8080', '#FF00FF', '#0000FF', '#00CCFF',
                  '#CCFFFF', '#CCFFCC', '#99CCFF', '#CC99FF', '#FFCC99', 
                  '#3366FF', '#33CCCC', '#99CC00', '#FF99CC', '#FFCC00'
      )
  
      currFactor = paste0("~", as.character(header[i]))
      plotPrefix = as.character(header[i])
  
      df = data.frame(sample_labels)
      colnames(df) = c("#SampleID")
      df = merge(df, mapping_file_parsed, by="#SampleID", all=FALSE, sort=FALSE)
      indice = nrow(alpha_div)
      selected_values = alpha_div[indice,]
      selected_values = t(selected_values)
      selected_values = data.frame(selected_values[1:nrow(selected_values),])
      names(selected_values) = c("diversity_value")
      selected_values$names = rownames(selected_values)
      selected_values$diversity_value[selected_values$diversity_value == "n/a"] <- 0
      selected_values$diversity_value[is.na(selected_values$diversity_value)] <- 0
      sv = merge(df, selected_values, by.x="#SampleID", by.y="names")
      colnames(sv)[1] = "SampleID"
      sv$diversity_value = as.numeric(sv$diversity_value)
      
      summary = summary(as.factor(sort(sv[[as.character(header[i])]])))
      meet_req = 1
      for(summaryEl in summary){
         if(summaryEl < 2){
            meet_req = 0
         }
      }
      if(meet_req == 1){
         ttest = pairwise.t.test(sv$diversity_value,g=factor(sv[[as.character(header[i])]]),paired=FALSE,pool.sd=FALSE)
      }else{
         #ttest = c(paste0("Can't perform test on ", currFactor, "factor: sample size too small", "x", "x"))
         ttest = NULL
         ttest$p.value = matrix(c("x","x","x","x"), nrow=2, ncol=2)
      }

      # Find number of samples to appropriately set font size.
      number_of_samples = (ncol(alpha_div) - 2)
      xFontSize = -0.0059 * number_of_samples + 6.1875
       
      p1 <- ggplot(environment=.e, data=sv, aes(x=SampleID, y=diversity_value)) +
      geom_bar(stat="identity", fill="#696969") +
      facet_grid(as.formula(currFactor), scales="free", space="free") + 
      theme(
         panel.border=element_rect(fill=NA, linetype="solid", colour = "black", size=1.4),
         axis.text.x=element_text(size=xFontSize, colour="black", angle=90), 
         axis.text.y=element_text(size=10, colour="black"),
         axis.title=element_text(size=10),
         axis.ticks.length=unit(0.2,"cm"),
         axis.ticks = element_line(colour = 'black', size = 0.5),
         panel.grid.minor=element_blank(),
         #panel.grid.major=element_line(colour="black", linetype="dotted", size=0.5),
         panel.grid.major=element_blank(),
         panel.background=element_blank(),
         legend.key.size = unit(0.45, "cm"),
         legend.margin = unit(1, "cm")
      ) + scale_y_continuous(limits=c(0,max(sv$diversity_value)))
  
      p2 <- ggplot(environment=.e, data=sv, aes(x=sv[[plotPrefix]], y=diversity_value, fill='gray')) +
      geom_boxplot() +
      stat_summary(fun.y=mean, geom="point", shape=5, size=4) +
      scale_fill_manual(values=c('gray')) +
      theme(
         panel.border=element_rect(fill=NA, linetype="solid", colour = "black", size=1.4),
         axis.text.x=element_text(size=10, colour="black", angle=90), 
         axis.text.y=element_text(size=10, colour="black"),
         axis.title=element_text(size=16),
         axis.ticks.length=unit(0.2,"cm"),
         axis.ticks = element_line(colour = 'black', size = 0.5),
         panel.grid.minor=element_blank(),
         panel.grid.major=element_line(colour="black", linetype="dotted", size=0.5),
         panel.background=element_blank(),
         legend.position="none"
         #legend.key.size = element_blank(),
         #legend.margin = element_blank()
      ) + 
      scale_y_continuous(limits=c(0,max(sv$diversity_value))) +
      annotation_custom(tableGrob(ttest$p.value), xmin=35, xmax=50, ymin=-2.5, ymax=-1) +
      xlab("Samples group")
 
      p3 <- ggplot(environment=.e, data=sv, aes(x=sv[[plotPrefix]], y=diversity_value)) +
      annotation_custom(tableGrob(ttest$p.value)) +
      theme(
         panel.grid.minor=element_blank(),
         panel.grid.major=element_blank(),
         axis.ticks = element_blank(),
         axis.text.y = element_blank(),
         axis.text.x = element_blank(),
         axis.title = element_blank(),
         panel.background=element_blank()
      )
  
      outfilePng = paste0(outdir, "/", prefix, "_", header[i], ".png")
      outfilePdf  = paste0(outdir, "/", prefix, "_", header[i], ".pdf")
  
      pdf( file=outfilePdf, height=20, width=16 )
      grid.arrange(p1, p2, p3, ncol=1, nrow=3)
      dev.off()

      options(bitmapType='cairo')
      png(file=outfilePng, height=12, width=10, units="in", res=300)
      grid.arrange(p1, p2, p3, ncol=1, nrow=3)
      dev.off()
   }   
 
	print("[DEBUG] Printed plots...")
} 

usage=function(errM) {
        cat("\nUsage : Rscript phylumBarplot.R [option] <Value>\n")
        cat("       -i        : infile\n")
        cat("       -o        : outdir\n")
        cat("       -p        : prefix\n")
        cat("       -m        : mapping file\n")
}

ARG = commandArgs(trailingOnly = T)

if(length(ARG) < 4) {
	usage("missing arguments")
}

## get arg variables
for (i in 1:length(ARG)) {
	if (ARG[i] == "-i") {
		infile=ARG[i+1]
	} else if (ARG[i] == "-o") {
		outdir=ARG[i+1]
	}else if (ARG[i] == "-p") {
		prefix=ARG[i+1]
	}else if(ARG[i] == "-m") {
      mappingFile=ARG[i+1]
   }
}

generateAlphaDivPlots(infile, outdir, prefix, mappingFile)

