#!/usr/bin/env perl

use strict;
use warnings;

use Getopt::Long;
use Iterator::FastaDb;
use File::Basename;
use Path::Class qw(dir);

my $usage=<<'ENDHERE';
NAME:
renameContigsMergedAnalysis.pl

PURPOSE:

INPUT:
--infile <string>  : fasta file (here generated by GAA).
--outfile <string> : fasta file output. 

OUTPUT:
Will write a new file having proper nomenclature for fasta headers.
>contig-<int> <int> nucleotides\n

NOTES:

BUGS/LIMITATIONS:
 
AUTHOR/SUPPORT:
 - Biomonitoring
Julien Tremblay - jtremblay514@gmail.com

ENDHERE

## OPTIONS
my ($help, $infile, $outfile);
my $verbose = 0;

GetOptions(
   'infile=s'   => \$infile,
   'outfile=s'  => \$outfile,
   'verbose' 	 => \$verbose,
   'help'       => \$help
);
if ($help) { print $usage; exit; }

## MAIN

open(OUT, ">".$outfile) or die "Can't open $outfile\n";

my $ref_fasta_db = Iterator::FastaDb->new($infile) or die("Unable to open Fasta file, $infile\n");
my $i = 1;
while( my $curr = $ref_fasta_db->next_seq() ) {
   my $header = $curr->header;
   my($nucleotides) = $header =~ /Contig\d+\.\d+\t(\d+)/;
   my $seq = $curr->seq;
   
   my $formattedSeq = "";
   while (my $chunk = substr($seq, 0, 150, "")) {
      $formattedSeq .= "$chunk\n";
   }
   print OUT ">contig-$i $nucleotides nucleotides\n";
   print OUT $formattedSeq."\n";
   $i++;
}
close(OUT);

